consensus/state.go

```diff
@@ -640,6 +640,8 @@ func (cs *ConsensusState) handleMsg(mi msgInfo) {
 	case *BlockPartMessage:
 		// if the proposal is complete, we'll enterPrevote or tryFinalizeCommit
 		_, err = cs.addProposalBlockPart(msg, peerID)
+		// TODO: At this point if added is true we have added new and useful block part so we can mark it in peer statistics.
```

Comment:
To be able to update statistics at this point we need access to PeerState. One way to achieve this would be to pass Peer instead of p2p.ID in msgInfo. We still have issue how to call MarkGood method of AddrBook from ConsensusState. One option would be having a channel that is read by Switch in which we can write MarkGood messages. Channel can be obtained either through Peer or directly from ConsensusState. Any other idea?